### ISSUE TITLE
[Android/F-Droid] F-Droid version online updating issue

### ISSUE THREAD

**Issue by guoyunhe** [libretro/RetroArch#17220]
## Description

Official website 1.19.1 APK is perfect on Android: it auto-download assets and you can download cores.

But F-Droid's 1.19.1 APK cannot auto-download assets and the core list is empty.

I guess F-Droid team mis-configured the build...

### Expected behavior

F-Droid APK should auto download assets and core list should not be empty. Just like the official APK:

![image](https://github.com/user-attachments/assets/8a72e598-0203-4e1d-9539-48a7c4f94da3)

### Actual behavior

F-Droid's 1.19.1 APK cannot auto-download assets and the core list is empty.

![image](https://github.com/user-attachments/assets/0c843608-6eca-4a68-8a42-e2c29051120a)

### Steps to reproduce the bug

1. Install APK from F-Droid
2. Launch the app
3. Go to Online Update

### Bisect Results

[Try to bisect and tell us when this started happening]

### Version/Commit

- RetroArch: 1.19.1

### Environment information

- OS: Android 14 (Xiaomi HyperOS 1.0)
- Compiler: [In case you are running local builds]

**Comment by hizzlekizzle** [comment]
did you try updating the info files first?

**Comment by guoyunhe** [comment]
> did you try updating the info files first?

Yes, I did. Actually, it forces you to update core info first before downloading cores...

**Comment by gouchi** [comment]
Even if you grant access to write external storage, it seems that default directory for cores and cores info is 

`/data/user/0/com.retroarch/cores`

If you switch to 

`/storage/emulated/0/RetroArch/cores`


It will work.

**Comment by guoyunhe** [comment]
> Even if you grant access to write external storage, it seems that default directory for cores and cores info is
> 
> `/data/user/0/com.retroarch/cores`
> 
> If you switch to
> 
> `/storage/emulated/0/RetroArch/cores`
> 
> It will work.

Is it a misconfigured build from F-Droid? Official APK works out-of-box, without any manual configuration.

**Comment by gouchi** [comment]
> Is it a misconfigured build from F-Droid? Official APK works out-of-box, without any manual configuration.

I don't know well Android but it does not seem they have changed something to the [build recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/06f169680c4829d912167de291788fd3d6a605a7/metadata/com.retroarch.yml#L192-L227).

**Comment by licaon-kter** [comment]
> I don't know well Android but it does not seem they have changed something to the [build recipe](https://gitlab.com/fdroid/fdroiddata/-/blob/06f169680c4829d912167de291788fd3d6a605a7/metadata/com.retroarch.yml#L192-L227).

true, and as the recipe is pretty short one can see we don't touch anything, afaik, so not sure that's what breaks it

HOME - Load Core - Download a Core - Fetch list... pops up... but stays empty, right @guoyunhe ?

All F-Droid versions do this, mmmkay

But...

HOME - Online Updater - touch Update CoreInfoFiles, then Assets, them the rest... Fetch list appears, files get downloaded, and put in place...

now, if you go to Download a Core... what do you see? :wink: 

anyway, https://github.com/libretro/RetroArch/issues/16126 should be fixed I guess

/PS: It does not help that there's no actual logcat error for us to grasp and find out why, even if the server is contacted, no content is pulled

**Comment by hizzlekizzle** [comment]
If this is an info/assets issue, would you want to switch to this alternate asset bundle, @licaon-kter ?

[glui_minimal_assets.zip](https://github.com/user-attachments/files/17996557/glui_minimal_assets.zip)

I know we talked about this and other strategies [in another issue](https://github.com/libretro/RetroArch/issues/16126#issuecomment-2015185317) but we never went any further with the other options, and this seems like a simple fix in the meantime.

**Comment by guoyunhe** [comment]
@licaon-kter exactly, F-Droid version seems missing some automatic download logic. You have to manually click everything in order by yourself. Official APK, on the other hand, is smarter and does all the essential downloading for you.

**Comment by licaon-kter** [comment]
@hizzlekizzle 
> would you want to switch to this alternate asset bundle,

switch to another one, yes

but I want it from http://buildbot.libretro.com/assets/frontend/ ... not a random Github link

**Comment by hizzlekizzle** [comment]
Sure. If you're prepared for that change, I'll get it up there. I just didn't want things to break by it suddenly having a different checksum.

**Comment by licaon-kter** [comment]
I can build a test one first, to see if it fixes stuff, brb

**Comment by licaon-kter** [comment]
> https://github.com/user-attachments/files/17996557/glui_minimal_assets.zip

using this one and the list is fetched right away :tada: 

when is the next version gonna be released?

**Comment by hizzlekizzle** [comment]
That's a good question. I don't have an ETA (I don't think anyone does, actually), but I can start talking to the other folks on the team about it. It's been a long time, so probably a good idea to get something going soon.

**Comment by guoyunhe** [comment]
> @hizzlekizzle
> 
> > would you want to switch to this alternate asset bundle,
> 
> switch to another one, yes
> 
> but I want it from http://buildbot.libretro.com/assets/frontend/ ... not a random Github link

+1, GitHub is blocked in China and some other countries. Using libretro domain & server is much safer.

**Comment by licaon-kter** [comment]
New update released but http://buildbot.libretro.com/assets/frontend/ -> glui_minimal_assets.zip is still from 2023?

**Comment by hizzlekizzle** [comment]
@licaon-kter sorry for the delay/hassle. I got someone to swap it out but he can't get to it until sometime tomorrow, but it should be good to go from that same buildbot address soon.

**Comment by hizzlekizzle** [comment]
@licaon-kter hey man, we should be good to go now. thanks for the heads-up, and thanks for working with us on it.

**Comment by licaon-kter** [comment]
added in https://gitlab.com/fdroid/fdroiddata/-/commit/f3244d34d1ff4de016d69b3b39e4c318e81c1ce5 thanks

**Comment by hizzlekizzle** [comment]
With that in place, are we ready to close this issue, @guoyunhe ?

**Comment by licaon-kter** [comment]
@guoyunhe https://gitlab.com/fdroid/wiki/-/wikis/FAQ#how-long-does-it-take-for-my-app-to-show-up-on-website-and-client

**Comment by licaon-kter** [comment]
did everyone got a chance to test 1.20?

**Comment by guoyunhe** [comment]
I just tested 1.20 from F-Droid and it still doesn't download assets etc. successfully...

**Comment by licaon-kter** [comment]
pics please, explain whet you did, what you wanted, what happened

**Comment by guoyunhe** [comment]
1. Uninstall RetroArch if installed old version.
2. Install RetroArch from F-Droid.
3. Open RetroArch and grant storage permission etc.

Actual behavior: It failed to download assets and cannot render Chinese fonts properly

![Image](https://github.com/user-attachments/assets/a7ba3bb6-c667-45fb-b9b3-93aa1e390d3c)

Expeced behavior: It should download assets and show Chinese fonts

**Comment by hizzlekizzle** [comment]
okay, does it resolve that if you include the fonts folder? https://github.com/libretro/retroarch-assets/tree/master/fonts

**Comment by guoyunhe** [comment]
> okay, does it resolve that if you include the fonts folder? https://github.com/libretro/retroarch-assets/tree/master/fonts

Chinese/Korean fonts are located at https://github.com/libretro/retroarch-assets/tree/master/pkg

I have tried official APK from https://retroarch.com/ and it downloads assets and fonts perfectly. But F-Droid version just doesn't work.

**Comment by gouchi** [comment]
@guoyunhe May you try to provide some log using something like [pidcat](https://repology.org/project/pidcat/versions) ?

`pidcat com.retroarch`

Thank you.

**Comment by matheuswillder** [comment]
> Chinese/Korean fonts are located at https://github.com/libretro/retroarch-assets/tree/master/pkg
> 
> I have tried official APK from https://retroarch.com/ and it downloads assets and fonts perfectly. But F-Droid version just doesn't work.

Just a quick heads-up here. The RetroArch build from the website doesn't download anything, *it extracts*. All the content is blunded in the APK (assets, core info files, shaders, etc.) and, on the first launch, it's extracted to the corresponding folders.

In the retroarch.cfg file, this is set with:

bundle_assets_extract_enable = "true", which is "true" on Android and "false" on other platforms.

This entry is also left unchanged in the build from F-Droid, but since the glui_minimal_assets.zip file doesn't include all those assets files, this issue is happening here.

So you need to download them manually if you want to use other fonts, shaders, or other menu driver (Ozone works surprisingly well with touchscreens), or more missing assets should be included in glui_minimal_assets.zip (F-Droid doesn't bundle all assets in the APK because the file hashes change/update frequently or something like this, if I remember correctly).

By the way, all these assets files are quite slow to extract on older devices, not because of the file size, but because of the sheer amount of files, more than 5.000 used for the XMB menu driver alone, that end up practically unused, since XMB isn't very touchscreen-friendly.

So I'm glad that glui_minimal_assets.zip exists now thanks to the F-Droid build.

**Comment by matheuswillder** [comment]
I'll post this in another comment instead of editing the previous one, as it's unrelated and just a suggestion.

@licaon-kter, according to this comment https://github.com/libretro/RetroArch/issues/11922#issuecomment-1510015356, the main assets.zip file update too fast and therefore cannot generate a checksum to satisfy F-Droid requirement (so that the build is reproducible, I think?).

The glui_minimal_assets.zip file was created for this reason and is currently used as source for the assets for the build in F-Droid, but it is not complete and contains only some essential files.

~I think this could be easily solved by pulling the assets from the latest stable Windows build instead of the main assets.zip file. It's just a .7z file, the contents of the /assets, /autoconfig, /database, /info, /overlays and /shaders folders are the same and will work (the other folders and files, such as the main .exe file, simply need to be ignored/deleted).~ Edit: it needs to be from the APK, the Windows build is still missing some assets used by Android, as I mentioned in a comment below.

And it can also be pulled from the APK from the website, they are inside the "assets" folder within the APK file, which would make the F-Droid build have the same assets as the website version.

Once the stable version is released, the checksum of the file and assets inside it will not change, thus resolving the issue and allowing the build in F-Droid to contain all the assets files.

The only issue I see is that it would be necessary to change to the new URL to pull the assets with each new version released, since currently the buildbot doesn't have a static URL that points always to the latest one.

Just my two cents.

**Comment by licaon-kter** [comment]
Merge requests welcomed :)

> (so that the build is reproducible, I think?).

no, so we make sure we pull the right assets

**Comment by matheuswillder** [comment]
>Merge requests welcomed :)

I'm really sorry, but I'm not a developer nor skilled with this, so I won't try to do it as it would only take your time or the time from someone on the team instead of really helping. I'm just a curious user lol

But if that's of any help, I was just now searching how to extract the assets from RetroArch.7z (Windows build) and RetroArch.apk (Android build from the website) using **7z** on Linux, and these do the trick:

`7z x "RetroArch.7z" RetroArch-Win64/{assets,autoconfig,database,filters,info,overlays,shaders}`

`7z x "RetroArch.apk" assets`

Other tools like **unzip** will probably be able to do the same too (extract only the folders that matter and ignore the rest), but I'm only familiar with 7z.

And the folder structure of the glui_minimal_assets.zip file is the same (assets, autoconfig, etc.)

Edit: some assets are missing between the Windows and Android builds, such /autoconfig/android folder on the Windows build (responsible for automatically recognizing/configuring the controls on Android). Therefore, the better source for extracting the assets would be from the APK.

Interestingly, apparently when you use the Online Updater, all assets are downloaded and become identical on both platforms, that's why I had previously assumed they were already the same.

**Comment by davidhedlund** [comment]
The F-Droid release does not ship with the assets for `armeabi-v7a` (32-bit RetroArch). The menu assets are missing so icons are missing and the font is severely pixelated. You have to run `Main Menu → Online Updater → Update Assets`.

Status for Google TV Streamer with Google TV 14:
* RetroArch 1.19.1: After you have run `Main Menu → Online Updater → Update Assets` you need to restart RetroArch to make load the menu assets the first time.
* RetroArch 1.20.0 - 1.22.2: After you have run `Main Menu → Online Updater → Update Assets` RetroArch automatically reloads the menu assets.

**Comment by davidhedlund** [comment]
@guoyunhe So can you verify if this issue was fixed in 1.20.0 for you too?

**Comment by guoyunhe** [comment]
@davidhedlund Tested in 1.22.2 (F-Droid):

1. Core info is downloaded automatically when I enter the core download list. 👍 
2. Assets are still not automatically downloaded. When I launch the app, it cannot render Chinese fonts. I have to manually download assets through online updater.

<img width="575" height="1278" alt="Image" src="https://github.com/user-attachments/assets/2f14c1bc-1743-4e55-b461-7142c75701d2" />

**Comment by davidhedlund** [comment]
> [@davidhedlund](https://github.com/davidhedlund) Tested in 1.22.2 (F-Droid):
> 
>     1. Core info is downloaded automatically when I enter the core download list. 👍
> 
>     2. Assets are still not automatically downloaded. When I launch the app, it cannot render Chinese fonts. I have to manually download assets through online updater.

Thank you for taking your time and testing!

Could you please update the issue title to something more specific, such as ‘_Update Assets does not render Chinese fonts_’?

**Comment by gouchi** [comment]
I checked the assets provided by the archive [glui_minimal_assets.zip](https://buildbot.libretro.com/assets/frontend/glui_minimal_assets.zip) and there is only one font provided which is (assets/glui/font.ttf)

So as mentioned [here](https://github.com/libretro/RetroArch/issues/17220#issuecomment-2515134196) and [there](https://github.com/libretro/RetroArch/issues/11501#issuecomment-720166763), we might need to update font to handle CJK languages.

**Comment by davidhedlund** [comment]
> I checked the assets provided by the archive [glui_minimal_assets.zip](https://buildbot.libretro.com/assets/frontend/glui_minimal_assets.zip) and there is only one font provided which is (assets/glui/font.ttf)
> 
> So as mentioned [here](https://github.com/libretro/RetroArch/issues/17220#issuecomment-2515134196) and [there](https://github.com/libretro/RetroArch/issues/11501#issuecomment-720166763), we might need to update font to handle CJK languages.

Thanks a lot for spotting this.

**Comment by davidhedlund** [comment]
> [@davidhedlund](https://github.com/davidhedlund) Tested in 1.22.2 (F-Droid):
> 
>     1. Core info is downloaded automatically when I enter the core download list. 👍

Great, then the issue in the main post is partially solved.

> 
>     2. Assets are still not automatically downloaded. When I launch the app, it cannot render Chinese fonts. I have to manually download assets through online updater.

Can you please add "[Feature request]: Auto-download assets" to the title because it's not a bug
